quinn-proto/src/connection/mod.rs

```diff
@@ -3034,6 +3079,53 @@ impl Connection {
                 self.stats.frame_tx.handshake_done.saturating_add(1);
         }
 
+        // OBSERVED_ADDR
+        let mut send_observed_address =
+            |space_id: SpaceId,
+             buf: &mut Vec<u8>,
+             max_size: usize,
+             space: &mut PacketSpace,
+             sent: &mut SentFrames,
+             stats: &mut ConnectionStats,
+             skip_sent_check: bool| {
+                // should only be sent within Data space and only if allowed by extension
+                // negotiation
+                // send is also skipped if the path has already sent an observed address
+                let send_allowed = self
+                    .config
+                    .address_discovery_role
+                    .should_report(&self.peer_params.address_discovery_role);
+                let send_required =
+                    space.pending.observed_addr || !self.path.observed_addr_sent || skip_sent_check;
```

Comment:
If the path stored the last acknowledged observed address that was sent, would that simplify this logic here?  It could check if the current remote still matches and send it if different in that case.  Instead of having to rely on having the right state.

Though I'm not really sure that would give the correct behaviour overall.

Reply:
I could try to think about it but too things come to mind trying to be pragmatic. The skip already helps us to send even when we are not wondering about ack'd frames at all: when we send with path challenges and responses. On the other side, and -again- trying to be pragmatic, this kind of logic change that does not fix a bug comes a bit late to the game and I'd be wary of introducing a bug by going this route